### PR TITLE
Close #8598: Adds ActivityDelegate to support the GeckoView delegate

### LINKED ISSUE #8598: Implement `GeckoRuntime.ActivityDelegate`
This new `GeckoRuntime` API allows GeckoView to essentially call `Activity#startActivityForResult()` via the embedding app. We currently need this to hook up the Web Authentication API. There is an example implementation in the GeckoView example app.

https://bugzilla.mozilla.org/show_bug.cgi?id=1549418

### PR DESCRIPTION
**Pull request by jonalmeida**
Added the `ActivityDelegate` with GeckoEngine support which will be used with a `WebAuthnFeature` component in the future.

GeckoView API docs [here](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/GeckoRuntime.ActivityDelegate.html).

Close #8598

### Pull Request checklist

- [x] **Quality content**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt
Comment (pocmo) [inline components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:454]: How are apps going to implement `ActivityDelegate`? Will this need a reference to an `Activity`? Is this something we can add to sample browser?

Comment (jonalmeida) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:454]: > How are apps going to implement `ActivityDelegate`? Will this need a reference to an `Activity`? 

I imagined the lifecycle for the activity delegate to be managed through the feature that implements it, but now I see that I would also need to introduce an unregister on the engine to notify it that we no longer have a delegate when we destroy from an activity or de-attach from a fragment.

> Is this something we can add to sample browser?

We could add something like this code below to the sample browser but it would tell the engine that we support WebAuthn features but our sample app cannot complete the request (needs to be an official signed app in the play store that is approved by Google to talk to their auth client).

```kotlin
// BaseBrowserFragment.kt
fun onStart() {
  engine.registerActivityDelegate(activityDelegate)
}

fun onStop() {
  engine.unregisterActivityDelegate()
}
```

We do have Reference Browser approved, so we can test this out in there before landing it in Fenix.

Comment (pocmo) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:454]: This still feels somewhat fragile: We have one engine, but potentially multiple activities / fragments displaying web content. How do we know which one to delegate this to? What to do in cases where we do not have an activity?

I wish the GV team would have added that to `GeckoView`, the thing that is actually bound to an Activity/Fragment.

Also: The API has somewhat unexpected side effects. The activity delegate pretends to be generic, but it seems to enable AuthN as a side effect? I do not understand how this is supposed to work as a generic delegate if there will be other reason for this delegate in the future.

Comment (jonalmeida) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:454]: > We have one engine, but potentially multiple activities / fragments displaying web content. How do we know which one to delegate this to? What to do in cases where we do not have an activity?

It would probably be the activity that the active GeckoSession is attached to with that engine. When we do not have an activity, do you mean the activity is destroyed? I've added an `unregister` method as well to remove the delegate in that case. If we're on a fragment, we could forward the activity events to the fragment in a similar way to [`UserInteractionHandler`](https://github.com/mozilla-mobile/android-components/blob/master/components/support/base/src/main/java/mozilla/components/support/base/feature/UserInteractionHandler.kt).

> I wish the GV team would have added that to GeckoView, the thing that is actually bound to an Activity/Fragment.

That's fair, I can bring this up with them.

> Also: The API has somewhat unexpected side effects. The activity delegate pretends to be generic, but it seems to enable AuthN as a side effect? I do not understand how this is supposed to work as a generic delegate if there will be other reason for this delegate in the future.

Good point, filed https://bugzilla.mozilla.org/show_bug.cgi?id=1671988

Comment (jonalmeida) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:454]: I opened a separate issue for the request to move the `ActivityDelegate` from the engine to the view: https://bugzilla.mozilla.org/show_bug.cgi?id=1672195

**Comment by jonalmeida** [comment]
@pocmo looking at the GV bugs filed as P3, should we move this forward for now?

In one of the bugs, it looks like this was intentional to have the ActivityDelegate on the Engine:

> On a related note, could the ActivityDelegate be added to the GeckoView instance instead where we have a reference to a fragment/activity? I can also file a separate issue for this discussion.
> 
> > It's possible, but it's not really in line with how Web Authn works in Gecko right now, so we punted on that in order to get things going sooner.

**Comment by pocmo** [comment]
To summarize my main concerns:

* The lifecycle of the activity and engine are not the same. So there's a risk of leaking the activity. We have a similar situation with the text selection delegate and that turned out to be fragile - even though we tried to register/unregister.

* There's not a guaranteed 1:1 mapping: There can be multiple activities (PWA, custom tabs) and there can be multiple sessions (loading in the background etc.). With that setup whatever activity registered last will handle the callback. How do we feel about initiating a start activity for a background tab, or a custom tab doing it for a tab that loaded in the browser?

@snorp Do you think that would be problematic?
@csadilek What do you think?

**Comment by snorp** [comment]
> To summarize my main concerns:
> 
>     * The lifecycle of the activity and engine are not the same. So there's a risk of leaking the activity. We have a similar situation with the text selection delegate and that turned out to be fragile - even though we tried to register/unregister.

Right. I recommend tracking the current Activity using the Lifecycle component. Keep a `WeakReference` to the current Activity in order to avoid accidental retainment.

> 
>     * There's not a guaranteed 1:1 mapping: There can be multiple activities (PWA, custom tabs) and there can be multiple sessions (loading in the background etc.). With that setup whatever activity registered last will handle the callback. How do we feel about initiating a start activity for a background tab, or a custom tab doing it for a tab that loaded in the browser?

There's only one "current" Activity, though, right? So a Custom Tab is fine if it's the current Activity.

Obviously this API is not ideal. We'd like to clean it up to at least operate on a session, but I thought it would be better to get this going sooner rather than later.

**Comment by snorp** [comment]
Hmm, it appears the lifecycle component doesn't seem to actually give you `Activity` references. That's unfortunate. I guess it has to happen from an `Application`.

**Comment by pocmo** [comment]
> There's only one "current" Activity, though, right? So a Custom Tab is fine if it's the current Activity.

* On at least Chrome OS there can be multiple resumed activities.
* What if the "current" activity (e.g. a custom tab) doesn't display the session that initiated the request though (e.g. a tab loaded in the background)?

> Obviously this API is not ideal. We'd like to clean it up to at least operate on a session, but I thought it would be better to get this going sooner rather than later.

This is fair. Let's unblock this.

**Comment by pocmo** [comment]
> Keep a WeakReference to the current Activity in order to avoid accidental retainment.

@jonalmeida Is it possible to have such a safe-guard?

**Comment by jonalmeida** [comment]
>     * What if the "current" activity (e.g. a custom tab) doesn't display the session that initiated the request though (e.g. a tab loaded in the background)?

This was fixed on desktop [recently](https://bugzilla.mozilla.org/show_bug.cgi?id=1540309), and now the requests are per window (gecko session for us), so we would need the same for the GV API eventually.

> @jonalmeida Is it possible to have such a safe-guard?

Yeah, I think this should be possible and something I'd like to do in the "WebAuthnFeature". For the engine, I could check if there is already a delegate registered and throw if a second one tries to do the same. For now, this could avoid instances where multiple features might try to register themselves.

I thought about keeping the `ActivityDelegate` within the engine and dispatching events to the `BrowserStore` as well, but that would cause web content to think we have support for WebAuthn because the delegate is attached (recall the other conversation thread).

What do you think?

**Comment by jonalmeida** [comment]
cc @pocmo if you feel strongly about this. I'll move ahead with my recommendation above.

> For the engine, I could check if there is already a delegate registered and throw if a second one tries to do the same. For now, this could avoid instances where multiple features might try to register themselves.

**Comment by pocmo** [comment]
I don't feel strongly about the case where we have multiple registered and the last one wins. My only concern would be accidental leaks - if we can prevent it. :)

**Comment by jmalinens** [comment]
I am eagerly waiting for this to work. We have implemented webauthn  in our website but I can not use webaurhn with my favorite mobile browser yet :/

**Comment by jonalmeida** [comment]
I've updated the PR to address the memory leak concerns and added a Fenix PR with an experimental WebAuthn feature would look like that we would like to land as well to test this with the privileged API from Play Services (which only worked on signed nightly builds).

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt
Comment (jonalmeida) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt:117]: @Amejia481 you may be interested in reviewing this change. 🙂

Comment (Amejia481) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt:117]: Sure, I will take a look today :)

Comment (Amejia481) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt:117]: Looks good ✅ !

**Review by csadilek — APPROVED** [review]
👓 Just some nits. My only concerns are also about activity leaks, but that is handled for now with the weak ref.

Can we link to the two GV issues in source maybe? I think it would help esp. https://bugzilla.mozilla.org/show_bug.cgi?id=1672195

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt
Comment (jonalmeida) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt:117]: Thanks!

**Comment by jonalmeida** [comment]
> Can we link to the two GV issues in source maybe? I think it would help esp. https://bugzilla.mozilla.org/show_bug.cgi?id=1672195

Done!

**Pull request merged**
